File: backend/app/services/indicative.py
"""Indicative fare estimation -- the resolution order that lets a
SerpApi-backed deployment (no "anywhere" endpoint) bootstrap flexible
destination discovery from cold start and improve as fare_observations
accumulates.

Resolution order:
  1. A fare_observations row for the exact (origin, destination, depart,
     return) no older than INDICATIVE_MAX_AGE_DAYS.
  2. The nearest-date observation(s) on the same route within
     +/-NEAREST_DATE_WINDOW_DAYS, distance-weighted.
  3. The static data/seed/route_baselines.json floor.

Tiers 1-2 are already origin-airport-specific (fare_observations rows are
keyed on the exact origin). Tier 3 is NOT -- route_baselines.json is keyed
by origin *metro* only, since Toronto-region airports share one metro. If
tier 3 ignored the airport entirely, every origin in the group would look
identical at cold start, which starves either the primary or the
alternates out of search/pruning.py's price-sorted survivor list before
they ever reach verification -- silently breaking the nearby-airport
savings feature (spec §31) on a fresh database. ORIGIN_FACTOR (shared
with the mock provider's simulated market) is applied so cold-start
estimates already reflect known airport-level pricing patterns.

All three tiers are also trip_type-scoped (round_trip vs one_way, see
providers/base.py) -- tiers 1-2 filter fare_observations by it so a
one-way price never answers a round-trip query for the same route/dates
or vice versa, and tier 3 applies ONE_WAY_FACTOR to the (round-trip-only)
baseline table. That flat factor is a known, accepted approximation: real
one-way/round-trip price ratios vary by route (competitive short-haul
routes often price near 50%, some long-haul/legacy-carrier routes well
above it), so a single constant can't capture that variance. It only
matters at cold start, though -- same multiplier applies to every date for
a given destination, so within-destination date ranking (what
search/pruning.py actually needs) is unaffected, and tiers 1-2 self-heal
as real one-way observations accumulate, same as ORIGIN_FACTOR already
does today.
"""

from __future__ import annotations

import datetime as dt
import json

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import SEED_DIR
from app.models.fares import FareObservation
from app.providers.base import TripType
from app.providers.mock import ORIGIN_FACTOR

INDICATIVE_MAX_AGE_DAYS = 7
NEAREST_DATE_WINDOW_DAYS = 10
GENERIC_FALLBACK_CAD = 1100.0
# Cold-start-only approximation applied to the round-trip baseline table --
# see module docstring.
ONE_WAY_FACTOR = 0.6

_BASELINES: dict[str, float] | None = None


def _baseline(origin_metro: str, destination: str) -> float | None:
    global _BASELINES
    if _BASELINES is None:
        with (SEED_DIR / "route_baselines.json").open("r", encoding="utf-8") as f:
            data = json.load(f)
        _BASELINES = data.get(origin_metro, {})
    return _BASELINES.get(destination)


async def estimate(
    session: AsyncSession,
    origin: str,
    destination: str,
    depart_date: dt.date,
    return_date: dt.date,
    *,
    trip_type: TripType = "round_trip",
    origin_metro: str = "toronto",
    now: dt.datetime | None = None,
) -> tuple[float, str]:
    """Returns (estimated_cad_price, source), source one of
    "observation_exact" | "observation_nearest" | "baseline".
    """
    now = now or dt.datetime.now(dt.UTC)

    cutoff = now - dt.timedelta(days=INDICATIVE_MAX_AGE_DAYS)
    exact = (
        (
            await session.execute(
                select(FareObservation)
                .where(
                    FareObservation.origin == origin,
                    FareObservation.destination == destination,
                    FareObservation.departure_date == depart_date,
                    FareObservation.return_date == return_date,
                    FareObservation.trip_type == trip_type,
                    FareObservation.observed_at >= cutoff,
                )
                .order_by(FareObservation.observed_at.desc())
            )
        )
        .scalars()
        .first()
    )
    if exact is not None:
        return exact.fare, "observation_exact"

    window_lo = depart_date - dt.timedelta(days=NEAREST_DATE_WINDOW_DAYS)
    window_hi = depart_date + dt.timedelta(days=NEAREST_DATE_WINDOW_DAYS)
    candidates = (
        (
            await session.execute(
                select(FareObservation).where(
                    FareObservation.origin == origin,
                    FareObservation.destination == destination,
                    FareObservation.trip_type == trip_type,
                    FareObservation.departure_date >= window_lo,
                    FareObservation.departure_date <= window_hi,
                )
            )
        )
        .scalars()
        .all()
    )
    if candidates:
        weighted_sum = 0.0
        weight_total = 0.0
        for obs in candidates:
            distance_days = abs((obs.departure_date - depart_date).days)
            weight = 1.0 / (1.0 + distance_days)
            weighted_sum += obs.fare * weight
            weight_total += weight
        if weight_total > 0:
            return round(weighted_sum / weight_total, 2), "observation_nearest"

    origin_factor = ORIGIN_FACTOR.get(origin, 1.0)
    baseline = _baseline(origin_metro, destination)
    base_price = baseline if baseline is not None else GENERIC_FALLBACK_CAD
    price = base_price * origin_factor
    if trip_type == "one_way":
        price *= ONE_WAY_FACTOR
    return round(price, 2), "baseline"

File: frontend/src/lib/engine/indicative.ts
// Indicative fare estimation used ONLY to prioritize which candidates are
// worth a live call -- never shown to the user as a price. Resolution
// order:
//   1. This user's own observation for the exact route/dates/party, <=7
//      days old.
//   2. Nearest-date observations on the same route+party within +/-10
//      days, distance-weighted.
//   3. No signal: candidates keep their original date-stride order
//      instead of being sorted by a fabricated number.
//
// There used to be a third tier here -- a static per-route baseline x
// origin factor. It's gone: with the origin/destination now able to be
// ANY airport in the world (not just the curated Toronto-area set the
// baseline table covered), a synthetic guess for an unseen route would be
// meaningless, and the user explicitly asked for live prices only, never
// guessed ones. Losing the tier only affects the ORDER candidates are
// tried in when this user has no history for the route -- every price
// actually shown always comes from a real SerpApi call (see
// verification.ts); this module never influences what's displayed.

import type { Passengers } from "@/lib/types";
import { loadObservations } from "./observations";
import type { EstimateSource } from "./types";

export const INDICATIVE_MAX_AGE_DAYS = 7;
export const NEAREST_DATE_WINDOW_DAYS = 10;

const DAY_MS = 86_400_000;

function daysApart(aIso: string, bIso: string): number {
  return Math.abs(Math.round((Date.parse(aIso) - Date.parse(bIso)) / DAY_MS));
}

/** Distinguishes a 2-adult observation from a 1-adult one so a party of
 * one can't be estimated from -- or poison the cache for -- a party of
 * four. Old rows saved before this field existed have no party_key; they
 * match only the common 1-adult/no-extras case, not any other party. */
export function partyKey(p: Passengers): string {
  return `${p.adults}-${p.children}-${p.infants_in_seat}-${p.infants_on_lap}`;
}

const SOLO_ADULT_KEY = partyKey({ adults: 1, children: 0, infants_in_seat: 0, infants_on_lap: 0 });

export function estimate(
  origin: string,
  destination: string,
  departDate: string,
  returnDate: string,
  tripType: string,
  passengers: Passengers,
  opts: { now?: number } = {},
): [number, EstimateSource] | [null, null] {
  const now = opts.now ?? Date.now();
  const key = partyKey(passengers);

  const rows = loadObservations().filter(
    (r) =>
      r.origin === origin &&
      r.destination === destination &&
      r.trip_type === tripType &&
      (r.party_key ?? SOLO_ADULT_KEY) === key,
  );

  const cutoff = now - INDICATIVE_MAX_AGE_DAYS * DAY_MS;
  const exact = rows
    .filter((r) => r.departure_date === departDate && r.return_date === returnDate && r.observed_at >= cutoff)
    .sort((a, b) => b.observed_at - a.observed_at)[0];
  if (exact) return [exact.fare, "observation_exact"];

  const nearby = rows.filter((r) => daysApart(r.departure_date, departDate) <= NEAREST_DATE_WINDOW_DAYS);
  if (nearby.length > 0) {
    let weightedSum = 0;
    let weightTotal = 0;
    for (const obs of nearby) {
      const weight = 1 / (1 + daysApart(obs.departure_date, departDate));
      weightedSum += obs.fare * weight;
      weightTotal += weight;
    }
    if (weightTotal > 0) {
      return [Math.round((weightedSum / weightTotal) * 100) / 100, "observation_nearest"];
    }
  }

  return [null, null];
}
